framework/filters/AccessRule.php

```diff
@@ -136,6 +142,10 @@ protected function matchController($controller)
      */
     protected function matchRole($user)
     {
+        if (!empty($this->permissions)) {
+            $this->roles = array_merge($this->roles, $this->permissions);
```

Comment:
I don't think we should affect the class state at all.
In case the rule runs more than once, the `roles` property will have N-as-many values.

```php
if (!empty($this->permissions)) {
    $roles = array_merge($this->roles, $this->permissions);
}
// then use $roles instead of $this->roles
```